<a href="https://www.kaggle.com/shivavashishtha/l12-replacing-missing-values?scriptVersionId=89779789" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/2022-ipl-auction-dataset/ipl_2022_dataset.csv


# Introduction 
There is no perfect way of replacing the null values, As it depends totally on the data and the analyst whi is cleaning the data. There are many operations in order to replace or fill the null values contained in a dataframe. However, the most common method is replacing the null values with the mean of that series. Let's try the same on this dataset

# Importing the Dataset

In [2]:
df=pd.read_csv("../input/2022-ipl-auction-dataset/ipl_2022_dataset.csv")

In [3]:
df

Unnamed: 0                  Player  Base Price         TYPE  \
0             0             Rashid Khan  Draft Pick       BOWLER   
1             1           Hardik Pandya  Draft Pick  ALL-ROUNDER   
2             2         Lockie Ferguson        2 Cr       BOWLER   
3             3           Rahul Tewatia     40 Lakh  ALL-ROUNDER   
4             4            Shubman Gill  Draft Pick       BATTER   
..          ...                     ...         ...          ...   
628         628            Sairaj Patil     20 Lakh       BATTER   
629         629              Monu Singh     20 Lakh       BOWLER   
630         630  Nivethan Radhakrishnan     20 Lakh       BOWLER   
631         631            Lance Morris     20 Lakh       BOWLER   
632         632            Aaron Hardie     20 Lakh  ALL-ROUNDER   

     COST IN ₹ (CR.)  Cost IN $ (000) 2021 Squad            Team  
0               15.0           1950.0        SRH  Gujarat Titans  
1               15.0           1950.0         MI  Gujarat Titans  
2               10.0           1300.0        KKR  Gujarat Titans  
3                9.0           1170.0         RR  Gujarat Titans  
4                8.0           1040.0        KKR  Gujarat Titans  
..               ...              ...        ...             ...  
628              NaN              NaN        NaN          Unsold  
629              NaN              NaN        NaN          Unsold  
630              NaN              NaN        NaN          Unsold  
631              NaN              NaN        NaN          Unsold  
632              NaN              NaN        NaN          Unsold  

[633 rows x 8 columns]

# Checking the info()

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 633 entries, 0 to 632
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       633 non-null    int64  
 1   Player           633 non-null    object 
 2   Base Price       633 non-null    object 
 3   TYPE             633 non-null    object 
 4   COST IN ₹ (CR.)  237 non-null    float64
 5   Cost IN $ (000)  237 non-null    float64
 6   2021 Squad       199 non-null    object 
 7   Team             633 non-null    object 
dtypes: float64(2), int64(1), object(5)
memory usage: 39.7+ KB


# Summing up the total null values present in the dataframe

In [5]:
df.isna().sum()

Unnamed: 0           0
Player               0
Base Price           0
TYPE                 0
COST IN ₹ (CR.)    396
Cost IN $ (000)    396
2021 Squad         434
Team                 0
dtype: int64

### Note:
We have 396 null values in column COST IN ₹ (CR.), however it is not suggested to use mean in order to fill the null values when the number is quite huge. But, just for the sake of learning let me teach you how to do it!

# Calculating the mean of the series by skipping null values

In [6]:
mean=df["COST IN ₹ (CR.)"].mean(skipna=True)

# Using the mean to fill the null values in the same series

In [7]:
df["COST IN ₹ (CR.)"]=df["COST IN ₹ (CR.)"].fillna(mean)

# Checking the sum of null values after filling in one of the series

In [8]:
df.isna().sum()

Unnamed: 0           0
Player               0
Base Price           0
TYPE                 0
COST IN ₹ (CR.)      0
Cost IN $ (000)    396
2021 Squad         434
Team                 0
dtype: int64

In [9]:
df

Unnamed: 0                  Player  Base Price         TYPE  \
0             0             Rashid Khan  Draft Pick       BOWLER   
1             1           Hardik Pandya  Draft Pick  ALL-ROUNDER   
2             2         Lockie Ferguson        2 Cr       BOWLER   
3             3           Rahul Tewatia     40 Lakh  ALL-ROUNDER   
4             4            Shubman Gill  Draft Pick       BATTER   
..          ...                     ...         ...          ...   
628         628            Sairaj Patil     20 Lakh       BATTER   
629         629              Monu Singh     20 Lakh       BOWLER   
630         630  Nivethan Radhakrishnan     20 Lakh       BOWLER   
631         631            Lance Morris     20 Lakh       BOWLER   
632         632            Aaron Hardie     20 Lakh  ALL-ROUNDER   

     COST IN ₹ (CR.)  Cost IN $ (000) 2021 Squad            Team  
0          15.000000           1950.0        SRH  Gujarat Titans  
1          15.000000           1950.0         MI  Gujarat Titans  
2          10.000000           1300.0        KKR  Gujarat Titans  
3           9.000000           1170.0         RR  Gujarat Titans  
4           8.000000           1040.0        KKR  Gujarat Titans  
..               ...              ...        ...             ...  
628         3.695781              NaN        NaN          Unsold  
629         3.695781              NaN        NaN          Unsold  
630         3.695781              NaN        NaN          Unsold  
631         3.695781              NaN        NaN          Unsold  
632         3.695781              NaN        NaN          Unsold  

[633 rows x 8 columns]

# Checking the occurrence of values using value_counts()

In [10]:
df["COST IN ₹ (CR.)"].value_counts()

3.695781     396
0.200000      52
0.500000      17
2.000000      13
4.000000       8
            ... 
11.500000      1
3.600000       1
12.250000      1
2.800000       1
3.200000       1
Name: COST IN ₹ (CR.), Length: 66, dtype: int64

# Filling a string in 2021 Squad Series
We have seen how to fill a float value inplace of null values, Now let's see if we could do the same for string values

In [11]:
df["2021 Squad"]=df["2021 Squad"].fillna("No Team")

In [12]:
df.isnull().sum()

Unnamed: 0           0
Player               0
Base Price           0
TYPE                 0
COST IN ₹ (CR.)      0
Cost IN $ (000)    396
2021 Squad           0
Team                 0
dtype: int64

In [13]:
df["2021 Squad"].value_counts()

No Team    434
SRH         26
MI          26
KKR         25
RR          25
CSK         25
DC          25
PBKS        24
RCB         23
Name: 2021 Squad, dtype: int64